src/main/java/org/squiddev/plethora/integration/astralsorcery/IntegrationAstralSorcery.java

```diff
@@ -0,0 +1,777 @@
+			out.put("knownConstellations", new LuaList<>(progress.getKnownConstellations()).asMap()); //FIXME ... great, it's the unlocalized name...
+
+			out.put("availablePerkPoints", progress.getAvailablePerkPoints(player)); //SO MANY METHODS THAT SHOULD BE STATIC, NOT INSTANCE!!!
+
+			IConstellation attuned = progress.getAttunedConstellation();
+			out.put("attunedConstellation", attuned != null ? context.makePartialChild(attuned).getMeta() : null);
+			out.put("progressTier", progress.getTierReached().toString()); //REFINE Do we want the name, the ordinal, or a LuaList with both?
+
+			// ... shouldn't the `progressId` field be the same as the ordinal? ... whatever.
+			out.put("researchTier", progress.getResearchProgression().stream()
+				.max(Comparator.comparingInt(ResearchProgression::getProgressId))
+				.<Object>map(Enum::toString).orElse(null));
+
+			//REFINE Someone else can expose the Perks if they want; cost/benefit says "no" at this time
+
+			return Collections.singletonMap("astralProgress", out);
+		}
+
+		@Nonnull
+		@Override
+		public EntityPlayer getExample() {
+			return new EntityPlayerDummy(WorldDummy.INSTANCE);
+		}
+	};
+
+	public static final IMetaProvider<IConstellation> META_I_CONSTELLATION = new BasicMetaProvider<IConstellation>() {
+
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IConstellation context) {
+			Map<String, Object> out = new HashMap<>(7);
+
+			out.put("unlocalizedName", context.getUnlocalizedName());
+			out.put("simpleName", context.getSimpleName());
+
+			out.put("color", context.getConstellationColor().getRGB()); //Used for particles on rituals, collectors?
+			out.put("colour", context.getConstellationColor().getRGB());
+			out.put("tierColor", context.getTierRenderColor().getRGB());
+			out.put("tierColour", context.getTierRenderColor().getRGB());
+
+			//Because there isn't a nice, clean `getTier` method...
+			out.put("tier", context instanceof IMinorConstellation ? "minor"
+				: context instanceof IMajorConstellation ? "major" : "weak");
+
+			//Exposing the stars and the connections wouldn't help players much, unless they want to visualize
+			// constellations on a computer...?
+
+			//Not wrapping in a namespace, as IConstellation is _usually_ a field on other objects
+			return out;
+		}
+
+		@Nullable
+		@Override
+		public IConstellation getExample() {
+			return getExampleConstellation();
+		}
+	};
+
+	public static final IMetaProvider<ItemStack> META_CONSTELLATION_PAPER = new ItemStackContextMetaProvider<ItemConstellationPaper>(
+		ItemConstellationPaper.class,
+		"Provides the Constellation, if any, for this Constellation Paper"
+	) {
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<ItemStack> context, @Nonnull ItemConstellationPaper item) {
+			IConstellation constellation = ItemConstellationPaper.getConstellation(context.getTarget());
+			return Collections.singletonMap("constellation",
```

Comment:
Probably better to do `constellation != null ? Collections.singletonMap(...) : Collections.emptyMap()`. It always feels a little weird creating maps with null values.

Reply:
Done; added external annotations to avoid this in the future.